src/backend/pipeline/cont_adhoc.c

```diff
@@ -0,0 +1,1000 @@
+	Tuplestorestate *batch = 0;
+	CleanupData cleanup_data;
+
+	ResourceOwner owner = 
+			ResourceOwnerCreate(CurrentResourceOwner, "WorkerResourceOwner");
+
+	acquire_cont_query_proc();
+
+	if (!MyContQueryProc)
+	{
+		elog(ERROR, "too many adhoc processes");
+	}
+
+	StartTransactionCommand();
+	CurrentResourceOwner = owner;
+	MemoryContextSwitchTo(run_cxt);
+	view_data = init_cont_view(stmt, s);
+	CommitTransactionCommand();
+
+	StartTransactionCommand();
+	CurrentResourceOwner = owner;
+	MemoryContextSwitchTo(run_cxt);
+	worker_receiver = CreateDestReceiver(DestTuplestore);
+
+	batch = tuplestore_begin_heap(true, true, 
+				continuous_query_combiner_work_mem);
+
+	SetTuplestoreDestReceiverParams(worker_receiver, batch,
+			CurrentMemoryContext, true);
+
+	worker_state = init_adhoc_worker(view_data, worker_receiver);
+	combiner_state = init_adhoc_combiner(view_data, batch);
+
+	if (combiner_state->existing)
+	{
+		keyColIdx = combiner_state->existing->keyColIdx;
+		numCols = combiner_state->existing->numCols;
+	}
+
+	adhoc_receiver = CreateAdhocDestReceiver(combiner_state->is_agg,
+										 keyColIdx,
+										 numCols);
+
+	view_state = init_adhoc_view(view_data, combiner_state->result,
+				 (DestReceiver *) adhoc_receiver);
+	view_state->slot = combiner_state->slot;
+
+	CommitTransactionCommand();
+
+	cleanup_data.cont_view_data = &view_data;
+	cleanup_data.worker_state = worker_state;
+
+	PG_ENSURE_ERROR_CLEANUP(cleanup, PointerGetDatum(&cleanup_data));
+	{
+		while (true)
+		{
+			bool new_rows = false;
+
+			/* use normal transaction machinery to cleanup any 'run' allocs */
+			StartTransactionCommand();
+			CurrentResourceOwner = owner;
+
+			new_rows = exec_adhoc_worker(worker_state);
+
+			if (new_rows)
+			{
+				exec_adhoc_combiner(combiner_state);
+				exec_adhoc_view(view_state);
+			}
+			else
+			{
+				/* TODO - check deadline timeout */
+				AdhocDestReceiverHeartbeat(adhoc_receiver);
+			}
+
+			CommitTransactionCommand();
+		}
+	}
+	PG_END_ENSURE_ERROR_CLEANUP(cleanup, PointerGetDatum(&cleanup_data));
+}
+
+/* 
+ * grabs the sql out of a statement like 
+ * select pipeline_exec_adhoc_query('select avg(x::int) from stream');
+ */
+static const char *
+get_stmt_sql(Node *node)
+{
+	/* assumes IsAdhocQuery has returned true on node already. */
+
+	SelectStmt *stmt = (SelectStmt *) (node);
+	Node *first_node = 0;
+	ResTarget *target = 0;
+	FuncCall *func = 0;
+	A_Const *sql_arg = 0;
+
+	first_node = (Node *) linitial(stmt->targetList);
+	target = (ResTarget *) (first_node);
+	func = (FuncCall *) (target->val);
+	sql_arg = linitial(func->args);
+
+	return sql_arg->val.val.str;
+}
+
+/* set a lower process priority for adhoc backends */
+static void
+set_nice_priority()
+{
+	int default_priority = getpriority(PRIO_PROCESS, MyProcPid);
+	int priority = Max(default_priority, 20 - ceil(continuous_query_proc_priority * (20 - default_priority)));
+
+	priority = nice(priority);
+}
+
+static void
+set_default_priority()
```

Comment:
Store the default priority as a static variable before calling `set_nice_property`? Don't think it's too important to reset it, since this is probably called when we're exiting from the client?